### PR TITLE
Xcode14 support

### PR DESCRIPTION
**Pull request by AlexeyBudynkov**
In Xcode 14 XCTAutomationSupport framework became impossible to link. So, some changes were required.

- removed XCTAutomationSupport linking, copying and signing
- Scanned Private Headers
- Commented unused dependencies
- Added changes after XCTAutomationSupport removal

### CODE REVIEW

File: Server/Application/Application.h [hunk 1]
```diff
@@ -6,6 +6,8 @@
 #import <UIKit/UIApplication.h>
 #import "XCTest+CBXAdditions.h"
 
+#import "FBXCElementSnapshot.h"
```
Comment (DmitriyKirakosyan) [inline Server/Application/Application.h]: Let's move this import to `.m` file.

File: Server/Application/Application.m [hunk 2]
```diff
@@ -192,16 +192,19 @@ + (void)launchAppWithBundleId:(NSString *_Nullable)bundleId
 + (NSDictionary *)tree {
     XCUIApplication *application = [Application currentApplication];
     XCUIElementQuery *applicationQuery = [XCUIApplication cbxQuery:application];
-    XCElementSnapshot *applicationSnaphot = [applicationQuery cbx_elementSnapshotForDebugDescription];
+//    /*XCElementSnapshot **/id applicationSnaphot = [applicationQuery cbx_elementSnapshotForDebugDescription];
```
Comment (DmitriyKirakosyan) [inline Server/Application/Application.m]: Remove commented lines.

File: Server/Application/Application.m [hunk 3]
```diff
@@ -192,16 +192,19 @@ + (void)launchAppWithBundleId:(NSString *_Nullable)bundleId
...
-+ (NSDictionary *)snapshotTree:(XCElementSnapshot *)snapshot {
+//+ (NSDictionary *)snapshotTree:(XCElementSnapshot *id )snapshot {
++ (NSDictionary *)snapshotTree:(id<FBXCElementSnapshot>)snapshot {
     NSMutableDictionary *json = [[JSONUtils snapshotOrElementToJSON:snapshot] mutableCopy];
 
     if (snapshot.children.count) {
-        NSMutableArray *children = [NSMutableArray array];
-        for (XCElementSnapshot *child in snapshot.children) {
+        NSMutableArray<id<FBXCElementSnapshot>> *children = [NSMutableArray array];
+        for (id child in snapshot.children) {
```
Comment (DmitriyKirakosyan) [inline Server/Application/Application.m]:
```suggestion
        for (id<FBXCElementSnapshot> child in snapshot.children) {
```

File: Server/Application/SpringBoard.m [hunk 4]
```diff
@@ -13,7 +13,7 @@
 #import "Application.h"
 #import "SpringBoardAlert.h"
 #import "SpringBoardAlerts.h"
-#import "XCElementSnapshot.h"
+//#import "XCElementSnapshot.h"
```
Comment (DmitriyKirakosyan) [inline Server/Application/SpringBoard.m]: Remove if not needed anymore.

File: Server/FBWebDriverAgent/FBElement.h [hunk 5]
```diff
@@ -0,0 +1,77 @@
+/**
```
Comment (DmitriyKirakosyan) [inline Server/FBWebDriverAgent/FBElement.h]: why don't we integrate it as a submodule?

File: Server/NSXPCConnection.h [hunk 6]
```diff
@@ -57,7 +57,7 @@
 @end
 
 @class XCSynthesizedEventRecord;
-@class XCElementSnapshot;
+//@class XCElementSnapshot;
```
Comment (DmitriyKirakosyan) [inline Server/NSXPCConnection.h]: remove

File: Server/PrivateHeaders/CBX-XCTest-Umbrella.h [hunk 7]
```diff
@@ -193,74 +183,10 @@
...
-#import "XCTestCore/XCUIScreen.h"
-#import "XCTestCore/XCUIScreenDataSource-Protocol.h"
-#import "XCTestCore/XCUIScreenshot.h"
-#import "XCTestCore/XCUIScreenshotProviding-Protocol.h"
-#import "XCTestCore/XCUISiriService.h"
-#import "XCTestCore/XCUISnapshotGenerationTracker.h"
-#import "XCTestCore/XCUITransformParameters.h"
 #import "XCTestCore/XCUIXcodeApplicationManaging-Protocol.h"
 #import "XCTestCore/_Dummy_ConformingToLegacyProtocolNames.h"
+//#import "XCTestCore/_XCTExceptionPointer.h"
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/CBX-XCTest-Umbrella.h]: remove

File: Server/PrivateHeaders/CDStructures.h [hunk 8]
```diff
@@ -11,8 +11,6 @@
 
 typedef void (^CDUnknownBlockType)(void); // return type and parameters are unknown
 
-typedef void (^CDUnknownFunctionPointerType)(void); // return type and parameters are unknown
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/CDStructures.h:14]: why removed? no alternative? why using `CDUnknownBlockType` instead is a correct approach?

File: Server/PrivateHeaders/DTXConnectionServices/DTXMessenger-Protocol.h [hunk 9]
```diff
@@ -12,6 +12,8 @@
 @protocol OS_dispatch_queue;
 @protocol OS_xpc_object;
 
+//#import <DTXConnectionServices/NSObject-Protocol.h>
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/DTXConnectionServices/DTXMessenger-Protocol.h]: remove

File: Server/PrivateHeaders/XCTAutomationSupport/XCElementSnapshot.h [hunk 10]
```diff
@@ -93,28 +100,29 @@
 @property BOOL isTruncatedValue;
 @property(copy) NSString *label;
 @property(copy) XCTLocalizableStringInfo *localizableStringInfo;
-@property(readonly) XCElementSnapshot *menu;
-@property(readonly) XCElementSnapshot *menuItem;
-@property(readonly) XCElementSnapshot *outline;
-@property __weak XCElementSnapshot *parent;
+@property(readonly) /*XCElementSnapshot **/id menu;
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/XCTAutomationSupport/XCElementSnapshot.h]: why?

File: Server/PrivateHeaders/XCTAutomationSupport/XCTApplicationStateSnapshot.h [hunk 11]
```diff
@@ -29,6 +30,7 @@
 @property(readonly) NSUInteger activationPolicy;
 @property(readonly, copy) NSString *bundleID;
 @property(readonly) NSUInteger eventID;
+@property(getter=isTrackingPID) BOOL trackingPID;
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/XCTAutomationSupport/XCTApplicationStateSnapshot.h:33]: readonly

File: Server/PrivateHeaders/XCTAutomationSupport/XCTApplicationStateSnapshot.h [hunk 12]
```diff
@@ -18,6 +18,7 @@
 
 @interface XCTApplicationStateSnapshot : NSObject <NSSecureCoding>
 {
+    BOOL _trackingPID;
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/XCTAutomationSupport/XCTApplicationStateSnapshot.h:21]: not needed

File: Server/PrivateHeaders/XCTAutomationSupport/XCTAutomationSession.h [hunk 13]
```diff
@@ -48,6 +49,7 @@
 @property(readonly) __weak id <XCTElementSnapshotProvider><XCTElementSnapshotAttributeDataSource> dataSource;
 @property(readonly) DTXConnection *dtxConnection;
 @property(readonly) BOOL hasBannerNotificationIsStickyAttribute;
+@property __weak id <XCTInternalEntitlementChecking> internalEntitlementChecker;
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/XCTAutomationSupport/XCTAutomationSession.h:52]: readonly

File: Server/PrivateHeaders/XCTAutomationSupport/XCTScreenshotRequest.h [hunk 14]
```diff
@@ -19,16 +19,19 @@
 
 @interface XCTScreenshotRequest : NSObject <NSSecureCoding>
 {
-    NSUInteger _screenID;
+    NSInteger _screenID;
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/XCTAutomationSupport/XCTScreenshotRequest.h:22]: _screenID can be negative?

File: Server/PrivateHeaders/XCTestCore/XCTMemoryMetric.h [hunk 15]
```diff
@@ -35,12 +34,10 @@
 @property(retain, nonatomic) XCTMemgraph *preGenerationMemgraph;
 @property(retain, nonatomic) NSString *processDisplayName;
 @property(retain, nonatomic) NSString *processIdentifierName;
-@property(retain, nonatomic) XCUIApplication *targetApplication;
-@property(retain, nonatomic) MXMMemoryMetric *underlyingMetric;
+@property(readonly, nonatomic) MXMMemoryMetric *underlyingMetric;
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/XCTestCore/XCTMemoryMetric.h:37]: any reason for this change?

File: Server/PrivateHeaders/XCTestCore/XCTMessagingRole_UIAutomationEventReporting-Protocol.h [hunk 16]
```diff
@@ -12,13 +12,13 @@
...
 
-@class NSArray, NSDictionary, NSNumber, NSString, XCElementSnapshot;
+@class NSArray, NSDictionary, NSNumber, NSString;
 
 @protocol XCTMessagingRole_UIAutomationEventReporting
-- (id)_XCT_nativeFocusItemDidChangeAtTime:(NSNumber *)arg1 parameterSnapshot:(XCElementSnapshot *)arg2 applicationSnapshot:(XCElementSnapshot *)arg3;
-- (id)_XCT_recordedEventNames:(NSArray *)arg1 timestamp:(NSNumber *)arg2 duration:(NSNumber *)arg3 startLocation:(NSDictionary *)arg4 startElementSnapshot:(XCElementSnapshot *)arg5 startApplicationSnapshot:(XCElementSnapshot *)arg6 endLocation:(NSDictionary *)arg7 endElementSnapshot:(XCElementSnapshot *)arg8 endApplicationSnapshot:(XCElementSnapshot *)arg9;
-- (id)_XCT_recordedFirstResponderChangedWithApplicationSnapshot:(XCElementSnapshot *)arg1;
-- (id)_XCT_recordedKeyEventsWithApplicationSnapshot:(XCElementSnapshot *)arg1 characters:(NSString *)arg2 charactersIgnoringModifiers:(NSString *)arg3 modifierFlags:(NSNumber *)arg4;
+- (id)_XCT_nativeFocusItemDidChangeAtTime:(NSNumber *)arg1 parameterSnapshot:(/*XCElementSnapshot **/id)arg2 applicationSnapshot:(/*XCElementSnapshot **/id)arg3;
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/XCTestCore/XCTMessagingRole_UIAutomationEventReporting-Protocol.h]: I don't like the idea of replacing `XCElementSnapshot` with `id`. Why did you do that?

File: Server/PrivateHeaders/XCTestCore/XCTRunnerDaemonSession.h [hunk 17]
```diff
@@ -14,132 +14,60 @@
...
 @property __weak id <XCTSignpostListener> signpostListener;
-@property(readonly) BOOL supportsAppearanceMode;
-@property(readonly) BOOL supportsAutomationMode;
-@property(readonly, nonatomic) BOOL supportsHEICImageEncoding;
-@property(readonly) BOOL supportsHIDEventRecording;
-@property(readonly) BOOL supportsInjectingVoiceRecognitionAudioInputPaths;
 @property(readonly) BOOL supportsPostingTelemetryData;
 @property(readonly) BOOL supportsSignpostListening;
-@property(readonly) BOOL supportsStartingSiriUIRequestWithAudioFileURL;
+@property __weak id <XCUIAXNotificationHandling> axNotificationHandler;
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/XCTestCore/XCTRunnerDaemonSession.h:43]: readonly?

File: Server/PrivateHeaders/XCTestCore/XCTRunnerDaemonSession.h [hunk 18]
```diff
@@ -14,132 +14,60 @@
...
-+ (void)legacyCapabilitiesForDaemonConnection:(id)arg1 completion:(CDUnknownBlockType)arg2;
-+ (void)modernCapabilitiesForDaemonConnection:(id)arg1 completion:(CDUnknownBlockType)arg2;
 + (void)sessionWithConnection:(id)arg1 completion:(CDUnknownBlockType)arg2;
 + (id)sharedSession;
 + (id)sharedSessionIfInitiated;
 + (id)sharedSessionPromiseAndImplicitlyInitiateSession:(BOOL)arg1;
-+ (id)unsupportedBundleIdentifiersForAutomationSessions;
-- (void)_XCT_applicationDidUpdateState:(id)arg1;
-- (void)_XCT_applicationWithBundleID:(id)arg1 didUpdatePID:(NSInteger)arg2 andState:(NSUInteger)arg3;
++ (id)testmanagerdServiceName;
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/XCTestCore/XCTRunnerDaemonSession.h:56]: no typo here?

File: Server/PrivateHeaders/XCTestCore/XCTRunnerIDESession.h [hunk 19]
```diff
@@ -43,8 +42,8 @@
 }
 
 @property(retain) XCTCapabilities *IDECapabilities;
-@property(retain) DTXConnection *IDEConnection;
-@property(retain) id <XCTMessagingChannel_RunnerToIDE, NSObject> IDEProxy;
+@property(readonly) DTXConnection *IDEConnection;
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/XCTestCore/XCTRunnerIDESession.h:45]: doesn't look like we need this change.

File: Server/PrivateHeaders/XCTestCore/XCTestExpectation.h [hunk 20]
```diff
@@ -29,13 +29,14 @@
...
     NSArray *_fulfillCallStackReturnAddresses;
-    NSUInteger _expectedFulfillmentCount;
     NSString *_expectationDescription;
+    NSUInteger _expectedFulfillmentCount;
     NSUInteger _numberOfFulfillments;
     NSUInteger _creationToken;
     NSArray *_creationCallStackReturnAddresses;
 }
 
+@property(copy, setter=_setExpectationDescription:) NSString *_expectationDescription;
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/XCTestCore/XCTestExpectation.h:39]: why underscore for property?

File: Server/PrivateHeaders/XCUIAutomation/XCUIScreen.h [hunk 21]
```diff
@@ -24,27 +24,30 @@
     BOOL _isMainScreen;
     XCUIKnobControl *_knobControl;
     XCUIButtonConsole *_buttonConsole;
-    NSInteger _displayID;
     id <XCUIDevice> _device;
     id <XCUIScreenDataSource> _screenDataSource;
+    NSInteger _displayID;
```
Comment (DmitriyKirakosyan) [inline Server/PrivateHeaders/XCUIAutomation/XCUIScreen.h:29]: Let's avoid unnecessary changes.

File: Server/XCTest+CBXAdditions.m [hunk 22]
```diff
@@ -123,7 +123,7 @@ - (void)cbx_resolve {
 
 @implementation XCUIElementQuery (CBXAdditions)
 
-- (XCElementSnapshot *)cbx_elementSnapshotForDebugDescription {
```
Comment (DmitriyKirakosyan) [inline Server/XCTest+CBXAdditions.m]: revert

File: Server/XCTest+CBXAdditions.h [hunk 23]
```diff
@@ -87,7 +86,7 @@
 
 @interface XCUIElementQuery (CBXAdditions)
 
-// Deprecated since Xcode 11.0
```
Comment (DmitriyKirakosyan) [inline Server/XCTest+CBXAdditions.h]: revert

File: Server/XCTest+CBXAdditions.h [hunk 24]
```diff
@@ -56,14 +55,14 @@
...
 - (XCUIElementQuery *_Nonnull)cbxQueryForDescendantsOfAnyType;
-- (XCElementSnapshot *_Nullable)cbxXCElementSnapshot;
+- (XCElementSnapshot* _Nullable)cbxXCElementSnapshot;
 + (void)cbxResolveApplication:(XCUIApplication *_Nonnull)xcuiApplication;
 
 @end
 
 @interface XCUIElement (CBXAdditions)
 
-- (XCElementSnapshot *_Nullable)lastSnapshot;
```
Comment (DmitriyKirakosyan) [inline Server/XCTest+CBXAdditions.h]: revert

Comment (DmitriyKirakosyan) [reply Server/XCTest+CBXAdditions.h]: and the ones above

File: Server/XCTest+CBXAdditions.h [hunk 25]
```diff
@@ -12,18 +12,17 @@
...
 #endif
 
+#import "XCUIHitPointResult.h"
 #import "CBXConstants.h"
 
-// Defined in XCTAutomationSupport framework, so it is safe to import
-// in any context.
 #import "XCElementSnapshot.h"
 
+
```
Comment (DmitriyKirakosyan) [inline Server/XCTest+CBXAdditions.h]:
```suggestion
```

File: Server/XCTest+CBXAdditions.h [hunk 26]
```diff
@@ -12,18 +12,17 @@
...
-#import "XCUIHitPointResult.h"
 #import "XCUIDevice.h"
 #import "XCUIApplicationStateTypedef.h"
 #import "XCTestCore/XCTest.h"
 #endif
 
+#import "XCUIHitPointResult.h"
 #import "CBXConstants.h"
 
-// Defined in XCTAutomationSupport framework, so it is safe to import
```
Comment (DmitriyKirakosyan) [inline Server/XCTest+CBXAdditions.h:23]: is this comment not relevant anymore?

File: Server/Utilities/JSONUtils.m [hunk 27]
```diff
@@ -76,11 +106,14 @@ + (NSDictionary *)snapshotOrElementToJSON:(id)element {
...
 
-        CBXVisibilityResult *result = [snapshot visibilityResult];
 
+        XCUIHitPointResult *hitPoint = [snapshot hitPoint:NULL];
+        CBXVisibilityResult *result = [self visibilityResult:hitPoint];
+        
         json[CBX_HITABLE_KEY] = @(result.isVisible);
         json[CBX_HIT_POINT_KEY] = @{@"x" : [JSONUtils normalizeFloat:result.point.x],
                                     @"y" : [JSONUtils normalizeFloat:result.point.y]};
+
```
Comment (DmitriyKirakosyan) [inline Server/Utilities/JSONUtils.m]:
```suggestion
```

File: Server/Utilities/JSONUtils.m [hunk 28]
```diff
@@ -37,11 +37,41 @@ + (NSArray *)elementTypes {
...
+}
+
 + (NSDictionary *)snapshotOrElementToJSON:(id)element {
     NSMutableDictionary *json = [NSMutableDictionary dictionary];
 
+    
     XCElementSnapshot *snapshot;
-    if ([element isKindOfClass:[XCElementSnapshot class]]) {
+    
+    //after Xcode 14 [XCElementSnapshot class] can't be get and this is the only way to compare classes
```
Comment (DmitriyKirakosyan) [inline Server/Utilities/JSONUtils.m]:
```suggestion
    //In Xcode 14 [XCElementSnapshot class] is not available, so this is the only way to compare classes.
```

**Comment by DmitriyKirakosyan** [comment]
@AlexeyBudynkov  Switch in TestApp doesn't work. Is it expected?

File: Server/Utilities/JSONUtils.m [hunk 29]
```diff
@@ -37,11 +37,41 @@ + (NSArray *)elementTypes {
             sortedArrayUsingSelector:@selector(localizedCaseInsensitiveCompare:)];
 }
 
++ (CBXVisibilityResult *)visibilityResult: (XCUIHitPointResult*)hitPoint {
+    CBXVisibilityResult *cbxResult = [CBXVisibilityResult new];
+    CGPoint intermediate;
+    XCUIHitPointResult *result = hitPoint;
```
Comment (IlyaBausovAkvelon) [inline Server/Utilities/JSONUtils.m:43]: Why do we need separate variable result? We already have an hitPoint variable which came from method parameters.

File: Server/Utilities/JSONUtils.m [hunk 30]
```diff
@@ -37,11 +37,41 @@ + (NSArray *)elementTypes {
...
+
 + (NSDictionary *)snapshotOrElementToJSON:(id)element {
     NSMutableDictionary *json = [NSMutableDictionary dictionary];
 
+    
     XCElementSnapshot *snapshot;
-    if ([element isKindOfClass:[XCElementSnapshot class]]) {
+    
+    //In Xcode 14 [XCElementSnapshot class] is not available, so this is the only way to compare classes.
+    Class targetClass = NSClassFromString(@"XCElementSnapshot");
```
Comment (IlyaBausovAkvelon) [inline Server/Utilities/JSONUtils.m:72]: Did you looked for XCElementSnapshot in other frameworks? Sometimes it can be just moved.

**Review by IlyaBausovAkvelon — CHANGES_REQUESTED** [review]

File: Server/Utilities/JSONUtils.m (hunk 29 above)
Comment (AlexeyBudynkov) [reply Server/Utilities/JSONUtils.m:43]: Previously this method was described in XCUIElement+VisibilityResult.m, but since XCElementSnapshot framework was hidden, I decided to implement this method here.

File: Server/Utilities/JSONUtils.m (hunk 30 above)
Comment (AlexeyBudynkov) [reply Server/Utilities/JSONUtils.m:72]: It wasn't moved, but its linking now doesn't work

**Comment by AlexeyBudynkov** [comment]
> @AlexeyBudynkov Switch in TestApp doesn't work. Is it expected?

Which one?

**Review by DmitriyKirakosyan — APPROVED** [review]

**Review by DmitriyKirakosyan — APPROVED** [review]

File: Server/Utilities/JSONUtils.m (hunk 30 above)
Comment (IlyaBausovAkvelon) [reply Server/Utilities/JSONUtils.m:72]: Ok

File: Server/Utilities/JSONUtils.m (hunk 29 above)
Comment (IlyaBausovAkvelon) [reply Server/Utilities/JSONUtils.m:43]: Ok

**Review by IlyaBausovAkvelon — APPROVED** [review]
